### PR TITLE
Add new gnss_* variables next to position_* variables

### PR DESCRIPTION
**Pull request by m-kuhn**
The position variables are only set when the digitizing crosshair is locked to the current positioning device location.
The gnss variables are always set.

The `position_source_name` will be set to `manual` when digitizing features with the position not locked

Fixes https://github.com/opengisch/QField/issues/1607

### CODE REVIEW

**Comment by nirvn** [comment]
Very good move.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for dd77126f49161203355e589458d1a524f094ae2d:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1608-99e8031f9aaa1f36e95124c2167457612842ab1f-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1608-99e8031f9aaa1f36e95124c2167457612842ab1f-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1608-99e8031f9aaa1f36e95124c2167457612842ab1f-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1608-99e8031f9aaa1f36e95124c2167457612842ab1f-x86.apk)

**Comment by LindaOPENGIS** [comment]
i tested @position_horizontal_accuracy in qfield. 
it worked properly: no values was returned when digitized a feature manually. a value was returned when using the gps position.

**Comment by m-kuhn** [comment]
@nirvn shall we backport this to 1.8 too?

**Comment by nirvn** [comment]
YesI think we should.

**Comment by signedav** [comment]
Do I understand correctly that these variables will only be available on collecting points with the cursor locked to the GNSS position? I see it's good reasons, but is there never the case where people want to collect polygons starting with the position then draw it and then use these values although it's not clear to what part of the geometry it fits? Just to be sure. I am not against this implementation. It will avoid erroneous data for sure.

**Comment by m-kuhn** [comment]
> is there never the case where people want to collect polygons starting with the position then draw it and then use these values although it's not clear to what part of the geometry it fits?

That's still possible with the values that are prefixed `source_` (e.g. `source_coordinate`, `source_horizontal_accuracy`, ...)

Thinking about it would probably be better to use `gnss_` or `sensor_` as prefix.

File: src/core/expressioncontextutils.cpp [hunk 1]
```diff
@@ -20,10 +20,21 @@
 #include "qgsgeometry.h"
 #include <QtPositioning/QGeoPositionInfoSource>
 
-QgsExpressionContextScope *ExpressionContextUtils::positionScope( const GnssPositionInformation &positionInformation )
+
+void addVar( QgsExpressionContextScope *scope, const QString &name, const QVariant &value, bool positionLocked, const QVariant &defaultValue = QVariant() )
+{
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "sensor_%1" ).arg( name ), value, true, true ) );
```
Comment (m-kuhn) [inline src/core/expressioncontextutils.cpp:26]:
```suggestion
  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "gnss_%1" ).arg( name ), value, true, true ) );
```

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for c3e0a24d7e61d05c1fe2d6daa957db5c0c3e4b7e:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1608-ac14ffb4c72a99b4b80f84e0a74d78a02bf96023-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1608-ac14ffb4c72a99b4b80f84e0a74d78a02bf96023-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1608-ac14ffb4c72a99b4b80f84e0a74d78a02bf96023-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1608-ac14ffb4c72a99b4b80f84e0a74d78a02bf96023-x86.apk)

File: src/core/expressioncontextutils.cpp [hunk 2]
```diff
@@ -20,10 +20,21 @@
 #include "qgsgeometry.h"
 #include <QtPositioning/QGeoPositionInfoSource>
 
-QgsExpressionContextScope *ExpressionContextUtils::positionScope( const GnssPositionInformation &positionInformation )
+
+void addVar( QgsExpressionContextScope *scope, const QString &name, const QVariant &value, bool positionLocked, const QVariant &defaultValue = QVariant() )
```
Comment (nirvn) [inline src/core/expressioncontextutils.cpp:24]: what about declaring the function _within_ ExpressionContextUtils::positionScope? The gnss,position variable prefixes below make this function only usable within that positionScope.

Alternative, rename to addPositionVariable?

**Review by nirvn — APPROVED** [review]
Looking good -- only one suggestion.

File: src/core/expressioncontextutils.cpp (hunk 2 above)
Comment (m-kuhn) [reply src/core/expressioncontextutils.cpp:24]: I thought it's easier to read without the lambda syntax, so let's go for `addPositionVariable`

Comment (m-kuhn) [reply src/core/expressioncontextutils.cpp:24]:
```suggestion
void addPositionVariable( QgsExpressionContextScope *scope, const QString &name, const QVariant &value, bool positionLocked, const QVariant &defaultValue = QVariant() )
```

File: src/core/expressioncontextutils.cpp [hunk 3]
```diff
@@ -43,27 +54,27 @@ QgsExpressionContextScope *ExpressionContextUtils::positionScope( const GnssPosi
...
+  addVar( scope, QStringLiteral( "vertical_accuracy" ), verticalAccuracy, positionLocked );
+
+  addVar( scope, QStringLiteral( "pdop" ), precisionDilution, positionLocked );
+  addVar( scope, QStringLiteral( "hdop" ), horizontalDilution, positionLocked );
+  addVar( scope, QStringLiteral( "vdop" ), verticalDilution, positionLocked );
+  addVar( scope, QStringLiteral( "number_of_used_satellites" ), numberOfUsedSatelites, positionLocked );
+  addVar( scope, QStringLiteral( "used_satellites" ),  QVariant::fromValue( usedSatelites ), positionLocked );
+  addVar( scope, QStringLiteral( "quality_description" ), qualityDescription, positionLocked );
+  addVar( scope, QStringLiteral( "fix_status_description" ), fixStatusDescription, positionLocked );
+  addVar( scope, QStringLiteral( "fix_mode" ), fixMode, positionLocked );
```
Comment (m-kuhn) [inline src/core/expressioncontextutils.cpp:77]:
```suggestion
  addPositionVariable( scope, QStringLiteral( "coordinate" ), QVariant::fromValue<QgsGeometry>( point ), positionLocked );
  addPositionVariable( scope, QStringLiteral( "timestamp" ), timestamp, positionLocked );
  addPositionVariable( scope, QStringLiteral( "direction" ), direction, positionLocked );
  addPositionVariable( scope, QStringLiteral( "ground_speed" ), groundSpeed, positionLocked );
  addPositionVariable( scope, QStringLiteral( "magnetic_variation" ), magneticVariation, positionLocked );
  addPositionVariable( scope, QStringLiteral( "horizontal_accuracy" ), horizontalAccuracy, positionLocked );
  addPositionVariable( scope, QStringLiteral( "vertical_accuracy" ), verticalAccuracy, positionLocked );
  addPositionVariable( scope, QStringLiteral( "3d_accuracy" ), horizontalVerticalAccuracy, positionLocked );
  addPositionVariable( scope, QStringLiteral( "vertical_speed" ), verticalSpeed, positionLocked );
  addPositionVariable( scope, QStringLiteral( "source_name" ), sourceName, positionLocked, QStringLiteral( "manual" ) );
  addPositionVariable( scope, QStringLiteral( "horizontal_accuracy" ), horizontalAccuracy, positionLocked );
  addPositionVariable( scope, QStringLiteral( "vertical_accuracy" ), verticalAccuracy, positionLocked );

  addPositionVariable( scope, QStringLiteral( "pdop" ), precisionDilution, positionLocked );
  addPositionVariable( scope, QStringLiteral( "hdop" ), horizontalDilution, positionLocked );
  addPositionVariable( scope, QStringLiteral( "vdop" ), verticalDilution, positionLocked );
  addPositionVariable( scope, QStringLiteral( "number_of_used_satellites" ), numberOfUsedSatelites, positionLocked );
  addPositionVariable( scope, QStringLiteral( "used_satellites" ),  QVariant::fromValue( usedSatelites ), positionLocked );
  addPositionVariable( scope, QStringLiteral( "quality_description" ), qualityDescription, positionLocked );
  addPositionVariable( scope, QStringLiteral( "fix_status_description" ), fixStatusDescription, positionLocked );
  addPositionVariable( scope, QStringLiteral( "fix_mode" ), fixMode, positionLocked );
```

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 7bfd04c8f8e1577e6ae4376cfc87ac7e97d35b46:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1608-da76ba82a3879aa3110ec59a980af781fb4784bc-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1608-da76ba82a3879aa3110ec59a980af781fb4784bc-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1608-da76ba82a3879aa3110ec59a980af781fb4784bc-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1608-da76ba82a3879aa3110ec59a980af781fb4784bc-x86.apk)

**Comment by nirvn** [comment]
Merge at will!

**Pull request merged**